### ISSUE TITLE
Reader highlighting / annotating is keyboard inaccessible

### ISSUE THREAD

**Issue by avaaec** [ManifoldScholar/manifold#2326]
This is a complex design issue with no clear precedent for how the keyboard interaction should behave, however we think the highlighting / annotating is a great feature and everyone should be able to use it!

## OBSERVED BEHAVIOR:
Manifold’s annotation feature is completely inaccessible by keyboard. There is no way to highlight the text, or bring up the options for annotation. Even if the options menu has been opened by a mouse, attempting to reach the menu using a keyboard dismisses it. As for previously made annotations, there is also no way to interact with those annotations using a keyboard from the text. In fact, they are not even announced as annotations, leading a non-visual user to perceive annotated text as simple plain text when reading.

## EXPECTED BEHAVIOR:
There is no true precedent for how this kind of annotation feature should interact. However, we have discussed ideas for the issue with many accessibility experts and developers, and have gathered a few ideas.

First, the text could be placed in a sort-of read-only form, content editable div, or text editing field. This would allow for a cursor, which would make highlight the text possible for keyboard users.

With this highlighting behavior, differing ideas were proposed for the pop-up menu. Some developers proposed that the menu would appear once shift was released, similar to how the menu would pop up when the mouse was released. The menu could then be dismissed completely using escape, or using shift to continue highlighting more text. 

There are definitely multiple ways to implement this in a accessible way.

## STEPS TO REPRODUCE:
1. Navigate to a published project (eg https://edge.manifoldapp.org/projects/the-black-tulip)
2. Navigate to the project’s main text, and move to a section with previous annotations (eg, the “A Grateful People” section of “The Black Tulip”)

### Mouse Functionality:
3. Move the mouse cursor over a portion of the text, and click and drag. Notice how the text is highlighted.
4. Release the mouse cursor, and notice how a context menu pops up for options (share, resource, annotate, highlight) for that selected text.
5. Attempt to reach this context menu with tab, and note how the menu is dismissed, making it unreachable.
5. Click on a previously annotated line of text, and note how there is a pop up window for interacting with that previous annotation. Note how there is no announcement for this popup window opening.

### Keyboard Functionality:
6. Read the text on the page with a screen reader by using arrow keys to navigate through the text. Note how text that has been highlighted or underlined as an annotation is not announced as an annotation, responding just like regular text.
7. Note how, if there are links in the text, tab can be used to navigate to those links. Note however, that tab cannot be used to navigate to the underlined text for previous annotations, despite the fact that they can be interacted with.
8. Attempt to highlight the text using keyboard keys: shift + arrow keys for individual characters, and shift + ctrl + arrow keys for full words. Note how, with no cursor, there is no indication for the user’s position.
9. Finally, note how there is no way to bring up the annotations options menu (share, resource, etc) by keyboard

## IMPLEMENTATION POC
```html
<div contenteditable oncut="return false"
  onpaste="return false"
  onkeydown="var key = event.keyCode; if(key == 37 || key == 39 || key == 9) return true; return false;">
  this is the text of the book yes this is the text of the book yes this is the text of the book yes this is the text of the book yes this is the text of the book yes this is the text of the book yes
</div>
```
* Until focused, the text looks like regular static text
* Allows keyboard users to navigate through the text easily with keyboard controls
* Allows keyboard users to highlight
* Just a proof of concept, still needs to allow for page up, page down, home, end, up arrow, down arrow and all other navigation keys, still needs popup functionality
* Not sure how this will work with rich text features like links, previous comments and annotations, resources. 
* Works in JsFiddle, not sure how browsers might handle this differently.

**Comment by zdavis** [comment]
Thanks for this!

I've done some exploration with this, and I do think it's possible to make the part of annotation that Manifold handles work without relying on mouse interactions. We mostly only use the mouse events to position the popup currently and are using other javascript interfaces to get the selected text. 

That said, there are two aspects to this:

1) The mechanics of selecting content
2) Everything that happens after the content is selected—responding to the selection, showing the popup, etc.

Currently, Manifold has no role in #1, nor should it. Selecting content on a webpage is a complex task, given that some kinds of content is not selectable. As such, the responsibility for how selection works is part of the functionality provided by the browser, not by the web application. Manifold comes in once the selection event has fired, and uses the browser's selection API to find out what text has been selected. There's really no way around that division of responsibilities, given the (sparse) interfaces that browsers expose around text selection.

I wonder how keyboard-only users select content on sites other than Manifold? For example, if you go to a page on nytimes.com or on washington.edu, is it possible to select a snippet of text without using the mouse? Adding a `contenteditable`, unfortunately, won't solve the problem in anything but the simplest applications, as there are many good reasons for content not to be editable and, moreover, keyboard users need to be able to select content that spans containers, just like mouse users can.

In my view, the problem here should definitely be addressed and taken seriously—it just has to happen with the browser vendors, as it's a global problem on the web.

I'm open to having my thinking here challenged. If you can point me toward an example of a complex web page with a wide range of HTML content that includes a mechanism for selecting text only using the keyboard, I will look at how they did it was done and see if we could utilize a similar approach. We might be able to craft an approach where a keyboard shortcut added tabindex to elements containing text nodes, so as to allow them to receive focus, and then another shortcut to make the node contenteditable for selection purposes. I'm not sure that the JS selection API would pick up on the selection in the same way, but we could explore this. We would need funding for this, I think, as getting this right would likely be rather involved.

**Comment by avaaec** [comment]
Thanks for such a thoughtful response! After more testing and research we have learned
* Most browsers offer a functionality for keyboard users to move a cursor arbitrarily around the page. Usually called caret browsing, this functionality allows keyboard users to highlight arbitrary text on a page. Right now in manifold the highlight / annotation popup does not appear when highlighting with caret browsing. It sounds like this is a pretty straight forward issue to fix. Additionally it's really easy to test, just enable caret browsing in Firefox with F7, or in chrome with the Caret browsing plugin and attempt to highlight text in the reader.
* Caret browsing however is not really compatible with more complex assistive technologies like screen readers. This makes the entire issue way more complicated. Essentially screen reader users don't have any ability to send a highlight event to the browser without the text being in the tab order or being "editable". We've done some research on possible solutions to this problem and have found that many digital text book interfaces basically make the static text focusable (with tabindex attribute) which then allows screen reader users to interact with it in a way were they can send highlight events. I would love to leave a link to one such example but the actual interface is usually hidden behind a login, we may however have an example we can share at a later date. I like your idea of adding a keyboard shortcut to add contenteditable / tabindex functionality to a particular spot in the text.
* I totally agree that the second part of this interaction, "Everything that happens after the content is selected" should be the same regardless of how the text is selected. If it works for keyboard users it should work fine for any most assistive tech users.

So to review:
* Making the functionality keyboard accessible should be pretty easy
* Making it work with assistive tech like screen readers will require more research, some of which we would be happy to help with.

**Comment by mkgold** [comment]
@zdavis We discussed this accessibility feature request on our call with Washington today. It is their highest priority needed bugfix. 

In talks with the team, we decided that the caret browsing solution would be a big step forward and would make Manifold annotation accessible for many users using assistive technology. Even though it wouldn't be a perfect solution, it would be a big step forward, so we agreed that we should do it.

As @avaaec described, caret browsing offers a way for users of assistive technology to select text, and the application can listen for an associated event (I think a lifting of the shift key, which is held down while selecting), to trigger the modal.

Can you please explore this and see whether it feasible?

**Comment by zdavis** [comment]
I'm pretty sure this would be doable, and perhaps not too big of a lift. I'll add it to our planned accessibility issues. I hear what you're saying about it being high priority, but it may come after other accessibility issues because it's more complex and probably requires a more senior developer to get it right.

**Comment by zdavis** [comment]
@bot bug

**Comment by zdavis** [comment]
Thank you for taking the time to report this bug. We've reviewed the issue and agree that this should be fixed. We’re adding a “planned” label to indicate that we consider this bug as part of our current work plan.

_This was an automated message, but please don't hesitate to reply. Our team watches these issues closely and will respond as soon as we're able to!_

**Comment by Santoz2** [comment]
Accessibility Triage Notes: UW's ATX team considers this a Deployment Blocker issue (Priority 1 Issue that would affect many users), as the primary functions of highlighting and annotating in the reader are not able to be performed using a keyboard.

**Comment by zdavis** [comment]
Support for keyboard annotation has been merged into master. Dana explained some of the details on the PR:

> This PR adds support for annotating texts via keyboard in the Manifold reader. There were two aspects to this work:
> 
> The first was adding the appropriate attributes to previously highlighted or annotated text nodes. Both now have a role of button, tabindex of 0, and an aria-label. They also received values of aria-haspopup appropriate to the expected interaction, with annotations getting dialog and highlights getting menu.
> 
> The second aspect was adding support for creating new annotations via keyboard. This work is scaffolded upon existing browser and OS functionality for keyboard text selection; @zdavis laid out on the relevant https://github.com/ManifoldScholar/manifold/issues/2326#issuecomment-523208990 why this is the responsibility of the browser and not Manifold. Our job was to ensure that text selection coming via keyboard triggered the annotation popup menu, which previously only responded to mouse and touch events. In addition, we needed to refactor the popup menu to be useable via keyboard and screen reader software. We've used the [WAI-ARIA menu design pattern](https://www.w3.org/TR/wai-aria-practices/#menu), with appropriate roles, focus management, and keyboard navigation via arrow keys.
> 
> To use this feature, users need to first enable caret browsing by pushing F7 in Firefox, Chrome, or Edge, or by [enabling Mouse Keys](https://support.apple.com/guide/mac-help/control-the-pointer-using-mouse-keys-mh27469/mac) in MacOS system accessibility preferences for use in Safari.

This feature works in Chrome, but there are still issues in Firefox and Safari. Both of those browser lose the text selection when the focus changes. We will do our best to solve this problem in non chromium browsers in the future, but for now, we're calling this issue complete given that we have support in Chromium browsers.
